### PR TITLE
[wasm][http] Improve compatibility of abort and cancelation of BrowserHttpHandler

### LINKED ISSUE #54270: [wasm][http] System.Net.Http.Functional.Tests.HttpClientTest - Expected: System.Threading.Tasks.TaskCanceledException
Unit tests:
- `GetStreamAsync_CanBeCanceled_AlreadyCanceledCts`
- `GetStringAsync_CanBeCanceled`
- `GetStreamAsync_CanBeCanceled`
- `GetByteArrayAsync_CanBeCanceled`
- `GetByteArrayAsync_CanBeCanceled_AlreadyCanceledCts`
- `GetStringAsync_CanBeCanceled_AlreadyCanceledCts`

```
Assert.Throws() Failure
Expected: typeof(System.Threading.Tasks.TaskCanceledException)
Actual:   typeof(System.OperationCanceledException): The operation was canceled.
   at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts) in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 618
   at System.Net.Http.HttpClient.GetStringAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken) in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 217
----- Inner Stack Trace -----
   at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts) in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 618
   at System.Net.Http.HttpClient.GetStringAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken) in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 217
```

Test `System.Net.Http.Functional.Tests.HttpClientTest.GetAsync_ContentCanBeCanceled`
```
 Assert.Throws() Failure
 Expected: typeof(System.Threading.Tasks.TaskCanceledException)
 Actual:   typeof(System.ObjectDisposedException): The CancellationTokenSource has been disposed.
 ---- System.ObjectDisposedException : The CancellationTokenSource has been disposed.
    at System.Threading.CancellationTokenSource.ThrowIfDisposed() in C:\Dev\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\CancellationTokenSource.cs:line 485
    at System.Threading.CancellationTokenSource.Cancel(Boolean throwOnFirstException) in C:\Dev\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\CancellationTokenSource.cs:line 249
    at System.Threading.CancellationTokenSource.Cancel() in C:\Dev\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\CancellationTokenSource.cs:line 221
    at System.Net.Http.BrowserHttpHandler.WasmFetchResponse.Dispose() in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\BrowserHttpHandler\BrowserHttpHandler.cs:line 369
    at System.Net.Http.BrowserHttpHandler.BrowserHttpContent.Dispose(Boolean disposing) in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\BrowserHttpHandler\BrowserHttpHandler.cs:line 434
    at System.Net.Http.HttpContent.Dispose() in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpContent.cs:line 689
    at System.Net.Http.HttpResponseMessage.Dispose(Boolean disposing) in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpResponseMessage.cs:line 235
    at System.Net.Http.HttpResponseMessage.Dispose() in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpResponseMessage.cs:line 242
    at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts) in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 583
    at System.Net.Http.HttpClient.GetStringAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken) in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 217
    at System.Net.Http.Functional.Tests.HttpClientTest.<>c__DisplayClass26_1.<<GetAsync_ContentCanBeCanceled>b__2>d.MoveNext() in C:\Dev\runtime\src\libraries\System.Net.Http\tests\FunctionalTests\HttpClientTest.cs:line 467
 --- End of stack trace from previous location ---
 ----- Inner Stack Trace -----
    at System.Threading.CancellationTokenSource.ThrowIfDisposed() in C:\Dev\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\CancellationTokenSource.cs:line 485
    at System.Threading.CancellationTokenSource.Cancel(Boolean throwOnFirstException) in C:\Dev\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\CancellationTokenSource.cs:line 249
    at System.Threading.CancellationTokenSource.Cancel() in C:\Dev\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\CancellationTokenSource.cs:line 221
    at System.Net.Http.BrowserHttpHandler.WasmFetchResponse.Dispose() in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\BrowserHttpHandler\BrowserHttpHandler.cs:line 369
    at System.Net.Http.BrowserHttpHandler.BrowserHttpContent.Dispose(Boolean disposing) in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\BrowserHttpHandler\BrowserHttpHandler.cs:line 434
    at System.Net.Http.HttpContent.Dispose() in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpContent.cs:line 689
    at System.Net.Http.HttpResponseMessage.Dispose(Boolean disposing) in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpResponseMessage.cs:line 235
    at System.Net.Http.HttpResponseMessage.Dispose() in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpResponseMessage.cs:line 242
    at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts) in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 583
    at System.Net.Http.HttpClient.GetStringAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken) in C:\Dev\runtime\src\libraries\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 217
    at System.Net.Http.Functional.Tests.HttpClientTest.<>c__DisplayClass26_1.<<GetAsync_ContentCanBeCanceled>b__2>d.MoveNext() in C:\Dev\runtime\src\libraries\System.Net.Http\tests\FunctionalTests\HttpClientTest.cs:line 467
```

### PR DESCRIPTION
**Pull request by pavelsavara**
- fixed handling of cancelation and abort exceptions to match unit test expectations
- for redirect outerloop tests added https://github.com/dotnet/runtime/issues/55083 

Fixes https://github.com/dotnet/runtime/issues/54270

### CODE REVIEW

**Review by ManickaP — APPROVED** [review]
LGTM, I'm not a browser handler expert though.

File: src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs [hunk 1]
```diff
@@ -482,10 +503,21 @@ public override async ValueTask<int> ReadAsync(Memory<byte> buffer, Cancellation
...
-                    catch (JSException)
+                    catch (OperationCanceledException oce ) when (cancellationToken.IsCancellationRequested)
+                    {
+                        throw CancellationHelper.CreateOperationCanceledException(oce, cancellationToken);
+                    }
+                    catch (JSException jse)
                     {
-                        cancellationToken.ThrowIfCancellationRequested();
-                        throw;
+                        if (jse.Message.StartsWith("AbortError"))
```
Comment (marek-safar) [inline src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:512]: Can the catch logic be refactored to a single method and reused?

File: src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs [hunk 2]
```diff
@@ -515,10 +547,21 @@ public override async ValueTask<int> ReadAsync(Memory<byte> buffer, Cancellation
...
-                catch (JSException)
+                catch (OperationCanceledException oce ) when (cancellationToken.IsCancellationRequested)
                 {
-                    cancellationToken.ThrowIfCancellationRequested();
-                    throw;
+                    throw CancellationHelper.CreateOperationCanceledException(oce, cancellationToken);
+                }
+                catch (JSException jse)
+                {
+                    if (jse.Message.StartsWith("AbortError"))
```
Comment (marek-safar) [inline src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:556]: I think all StartsWith would better use StringComparison.Ordinal

File: src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs [hunk 3]
```diff
@@ -323,9 +323,21 @@ protected internal override async Task<HttpResponseMessage> SendAsync(HttpReques
                 return httpResponse;
 
             }
-            catch (JSException jsExc)
+            catch (OperationCanceledException oce ) when (cancellationToken.IsCancellationRequested)
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:326]:
```suggestion
            catch (OperationCanceledException oce) when (cancellationToken.IsCancellationRequested)
```

**Comment by marek-safar** [comment]
Is this ready for merge?

**Review by lewing — APPROVED** [review]

**Comment by lewing** [comment]
Failing test in runtime (Libraries Test Run release mono Linux x64 Debug) is

```
 Starting:    System.Threading.Tests (parallel test collections = on, max threads = 2)
    System.Threading.Tests.ThreadLocalTests.RunThreadLocalTest8_Values_NegativeCases [SKIP]
      Condition(s) not met: "IsPreciseGcSupported"
    System.Threading.Tests.ThreadLocalTests.RunThreadLocalTest5_Dispose [FAIL]
      Assert.True() Failure
      Expected: True
      Actual:   False
      Stack Trace:
        /_/src/libraries/System.Threading/tests/ThreadLocalTests.cs(139,0): at System.Threading.Tests.ThreadLocalTests.RunThreadLocalTest5_Dispose()
        /_/src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs(370,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

**Pull request merged**